broker/src/main/java/io/zeebe/broker/system/partitions/impl/PartitionTransitionImpl.java

```diff
@@ -18,101 +19,78 @@
 import java.util.List;
 import org.slf4j.Logger;
 
-public class PartitionTransitionImpl<T> implements PartitionTransition {
+public class PartitionTransitionImpl implements PartitionTransition {
 
   private static final Logger LOG = Loggers.SYSTEM_LOGGER;
+  private static final List<PartitionStep> EMPTY_LIST = ImmutableList.of();
 
-  private final PartitionContext state;
-  private final List<Component<T>> leaderComponents;
-  private final List<Component<T>> followerComponents;
-  private final List<Component<T>> openedComponents = new ArrayList<>();
+  private final PartitionContext context;
+  private final List<PartitionStep> leaderSteps;
+  private final List<PartitionStep> followerSteps;
+  private final List<PartitionStep> openedSteps = new ArrayList<>();
+  private CompletableActorFuture<Void> currentTransition = CompletableActorFuture.completed(null);
 
   public PartitionTransitionImpl(
-      final PartitionContext state,
-      final List<Component<T>> leaderComponents,
-      final List<Component<T>> followerComponents) {
-    this.state = state;
-    this.leaderComponents = leaderComponents;
-    this.followerComponents = followerComponents;
+      final PartitionContext context,
+      final List<PartitionStep> leaderSteps,
+      final List<PartitionStep> followerSteps) {
+    this.context = context;
+    this.leaderSteps = leaderSteps;
+    this.followerSteps = followerSteps;
   }
 
   @Override
-  public void toFollower(final CompletableActorFuture<Void> future) {
-    closePartition()
-        .onComplete(
-            (nothing, err) -> {
-              if (err == null) {
-                installComponents(future, new ArrayList<>(followerComponents));
-              } else {
-                future.completeExceptionally(err);
-              }
-            });
+  public ActorFuture<Void> toFollower() {
+    final CompletableActorFuture<Void> future = new CompletableActorFuture<>();
+    currentTransition.onComplete((nothing, err) -> transition(future, followerSteps));
+    return future;
```

Comment:
This may lead to problems if two role changes happen before the current transition is done. Both transitions would be performed after the current transition is completed. But the second transition should wait until the previous transition is completed.

However, this is maybe not in the scope of this refactoring. So, feel free to ignore :sweat_smile:

Reply:
Very true. I won't fix it now since, like you said, it's out of scope but I'll make an issue for all the known issues of the ZeebePartition. Thanks!